src/realm/alloc.hpp

```diff
@@ -158,10 +158,36 @@ class Allocator {
     // The ref translation splits the full ref-space (both below and above baseline)
     // into equal chunks.
     struct RefTranslation {
-        char* mapping_addr;
+        char* mapping_addr = nullptr;
+        std::atomic<size_t> lowest_possible_xover_offset = 0;
+
+        // member 'xover_mapping_addr' is used for memory synchronization of the fields
+        // 'xover_mapping_base' and 'xover_encrypted_mapping'.
+        std::atomic<char*> xover_mapping_addr = nullptr;
+        size_t xover_mapping_base = 0;
 #if REALM_ENABLE_ENCRYPTION
-        util::EncryptedFileMapping* encrypted_mapping;
+        util::EncryptedFileMapping* encrypted_mapping = nullptr;
+        util::EncryptedFileMapping* xover_encrypted_mapping = nullptr;
 #endif
+        RefTranslation(char* addr)
+            : mapping_addr(addr)
+        {
+        }
+        RefTranslation() = default;
+        RefTranslation& operator=(const RefTranslation& from)
+        {
+            if (&from != this) {
+                mapping_addr = from.mapping_addr;
+                lowest_possible_xover_offset.store(from.lowest_possible_xover_offset, std::memory_order_relaxed);
+                xover_mapping_base = from.xover_mapping_base;
+#if REALM_ENABLE_ENCRYPTION
+                encrypted_mapping = from.encrypted_mapping;
+                xover_encrypted_mapping = from.xover_encrypted_mapping;
+#endif
+                xover_mapping_addr.store(from.xover_mapping_addr, std::memory_order_release);
```

Comment:
I think this is subtly incorrect since we are reading from `from.xover*` before the acquire-load from `from.xover_mapping_addr`. I think we also risk storing a stale `lowest_possible_xover_offset` wrt `xover_mapping_adder`. I think something like this would be more correct:

```cpp
        RefTranslation& operator=(const RefTranslation& from)
        {
            if (&from != this) {
                mapping_addr = from.mapping_addr;
#if REALM_ENABLE_ENCRYPTION
                encrypted_mapping = from.encrypted_mapping;
#endif

                const auto local_xover_mapping_addr = from.xover_mapping_addr.load(std::memory_order_acquire);

                // This must be loaded after xover_mapping_addr to ensure it isn't stale.
                lowest_possible_xover_offset.store(from.lowest_possible_xover_offset, std::memory_order_relaxed);

                if (local_xover_mapping_addr) {
                    xover_mapping_base = from.xover_mapping_base;
#if REALM_ENABLE_ENCRYPTION
                    xover_encrypted_mapping = from.xover_encrypted_mapping;
#endif
                    xover_mapping_addr.store(local_xover_mapping_addr, std::memory_order_release);
                }
            }
            return *this;
        }
```

This may be OK as written because all potential races here are covered by the `m_mapping_mutex` but relying on that seems a bit fragile. If you want to rely on that, then I think we can safely make this use relaxed load/store, and use `m_ref_translation_ptr` for memory ordering.

Reply:
I agree, will fix.